### DISCUSSION TITLE
ListObjectV2 response time slow

### DISCUSSION THREAD

**Discussion by grydan** [minio/minio#18115] [Q&A]
I have deplyed a MinIO server into my VM under ESXI with 4vCPU and 8GB of RAM and with Rocky Linux 9.2 that works well and has a dedicated storage area of 10 Tera.

Setup was trivial by following this guide:

https://min.io/docs/minio/linux/operations/install-deploy-manage/deploy-minio-single-node-single-drive.html#minio-snsd

I have another application server that has a native feature that allows me to move all files from local storage to MinIO, and so I enabled it...Consider that locally I have 10 million files that are moved to MinIO....

During this shift 3 S3 calls are invoked to MINIO ...
The first one is **S3:ListObjectV2** that as the number of files moved to MinIO increases latency response (I now have 1 million files in MinIO) it takes about 1 second to respond...and this slows down the data migration process.

Do you have any suggestions or best praticse to follow or optimize suggestion?

The cpu usage is around 30% , the memory consumed is 50% and I don't have any bottleneck in disk I/O.
 
I have run a trace trace with mc client the traffic and  I observed that for every files to push the application server run this three S3 call:

mrockyminio.testmachine9000 [REQUEST **s3.ListObjectsV2**] [2023-09-25T18:35:28.622] [Client IP: 172.29.0.201]
vmrockyminio.testmachine9000 GET /cldaossvil?list-type=2&prefix=ref%2F786BE78022D607E9F029B6AA340776B41BCD4661%2F0B72E7AD42892B14D6DA769FB2E5150BC592686200178E35
vmrockyminio.testmachine9000 Proto: HTTP/1.1
.
.
.
vmrockyminio.testmachine9000 [REQUEST **s3.HeadObject**] [2023-09-25T18:35:28.762] [Client IP: 172.29.0.201]
vmrockyminio.testmachine9000 HEAD /cldaossvil/obj/0000000000000000000000000000000000000000/786BE78022D607E9F029B6AA340776B41BCD4661/0B72E7AD42892B14D6DA769FB2E5150BC592686200178E35.nlo
.
.
.
vmrockyminio.testmachine9000 [REQUEST **s3.PutObject**] [2023-09-25T18:35:28.776] [Client IP: 172.29.0.201]
vmrockyminio.testmachine9000 PUT /cldaossvil/obj/0000000000000000000000000000000000000000/786BE78022D607E9F029B6AA340776B41BCD4661/0B72E7AD42892B14D6DA769FB2E5150BC592686200178E35.nlo

**Comment by klauspost** [comment]
https://min.io/docs/minio/kubernetes/upstream/administration/object-management.html#object-organization-and-planning

> Deployments with modest or budget-focused hardware should architect their workloads to target 10,000 objects per prefix as a baseline. Increase this target based on benchmarking and monitoring of real world workloads up to what the hardware can meaningfully handle.

"Prefix" is the `obj/0000000000000000000000000000000000000000/786BE78022D607E9F029B6AA340776B41BCD4661/` part of your key.

**Reply by grydan** [reply]
Can I modulate my hardware (is a enterprise hosts with a SAN) but i can't see any hardware problem : CPU, RAM and I/O disk is normal...

**Reply by klauspost** [reply]
We only offer performance optimization as part of our [subscriptions](https://min.io/pricing).

You are welcome to reach out on `sales@min.io` and describe your case and we can get you set up so we can look deeper into this.
